Infrastructure as Code. Terraform

Durf1985 opened the pull request:
## IaC Practice using Terraform in GCP


## In the process done:

* Install Terraform
* Create provider
* Create VM instance
* Creating Firewall rules
* Creating global metadata key for all instances
* Planing and apply terraform template
* Creating output variables file
* Using `reddit-base` image with provisioners, instead `reddit-full` image
* Input vars file
* Create load balancer with parametrization

## How to launch a project:

```
terraform init
terraform plan
terraform apply
```

## How to check the performance:

```bash
gcloud compute addresses describe lb-ipv4-1 \                                                                                 
   --format="get(address)" \
   --global
```
Paste the received IP address into the address bar of your browser. You have to get through the load balancer into your reddit application

## PR checklist

The label with the homework number, in label description - homework lecture and topic

Durf1985 committed:
DO-41 Create an instance via terraform
Added working main.tf template based on image "reddit-base" to create
VM instance with ssh access

Durf1985 commented:
@loktionovam created VM instance via terraform
![photo_2022-09-22_16-20-37](https://user-images.githubusercontent.com/36097814/191709554-cab5adb7-89ec-4568-b7af-dfbbea6fbd79.jpg)
![photo_2022-09-22_16-20-52](https://user-images.githubusercontent.com/36097814/191709562-f269d431-df6b-4b54-a0f8-2620498d12e7.jpg)

loktionovam commented on .gitignore:
```diff
@@ -1,3 +1,8 @@
+*.tfstate.*.backup
+*.tfstate.backup
+*.tfvars
+.terraform/
```
add `.terraform.lock.hcl` also

loktionovam commented on terraform/.terraform.lock.hcl:
```diff
@@ -0,0 +1,21 @@
+# This file is maintained automatically by "terraform init".
```
remove this file because it auto generated by `terraform init` command and is not a part of your source code

Durf1985 committed:
DO 45-46 Firewall and reddit app via terraform
Create firewall rules via terraform
Install reddit application via terraform provisioners

loktionovam commented on terraform/outputs.tf:
```diff
@@ -0,0 +1,3 @@
+output "app_external_ip" {
+  value = google_compute_instance.app.network_interface.0.access_config.0.nat_ip
```
Add the description of this output

https://cloud.google.com/docs/terraform/best-practices-for-terraform#outputs

loktionovam commented on terraform/main.tf:
```diff
@@ -0,0 +1,62 @@
+
+  }
+  metadata = {
+    ssh-keys = "appuser:${file("~/github_repo/devops_study_infra/packer/scripts/TXT.pub")}"
```
This absolute path can be different on different computer (for example my path will be `~/src/github/bla-bla`) so replace it by relative path

Durf1985 committed:
DO-47,48 Parametrize, public key metadata

Parametrize terraform configuration
Install public keys to project metadata

Durf1985 committed:
DO-50 Create a loadbalancer

file lb.tf - contains a raw load balancer project, nominally launched,
but has not passed normal testing, since the sandbox does not work.
Does not contain the normal names of resources and modules

main.tf - it was redesigned for the "reddit-full" image, in which
a unit was created and the application runs out of the box
instead of "reddit-base", which requires separate scripts in the
terraform configuration, and therefore an increase in the loading
time of the instance

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+    boot         = true
+    device_name  = "persistent-disk-0"
+    mode         = "READ_WRITE"
+    source_image = "projects/clgcporg2-118/global/images/reddit-full-1664802177" 
```
here you should use your project variable rather than hardcoding

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+      network_tier = "PREMIUM"
+    }
+    network    = "global/networks/default" 
+    subnetwork = "regions/us-central1/subnetworks/default"
```
use region variable there rather than hardcoding

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+resource "google_compute_instance_template" "default" {
```
replace `default` resource name by something more suitable

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+  tags = ["reddit-app"]
+}
+
+resource "google_compute_instance_group_manager" "default" {
```
replace `default` resource name by something more suitable

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+  target_size        = 2
+}
+
+resource "google_compute_firewall" "default" {
```
replace `default` resource name by something more suitable

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+  }
+}
+
+resource "google_compute_global_address" "default" {
```
replace `default` resource name by something more suitable

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+  ip_version = "IPV4"
+}
+
+resource "google_compute_health_check" "default" {
```
replace `default` resource name by something more suitable

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+  unhealthy_threshold = 2
+}
+
+resource "google_compute_backend_service" "default" {
```
replace `default` resource name by something more suitable

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+}
+
+
+resource "google_compute_url_map" "default" {
```
replace `default` resource name by something more suitable

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+  default_service = google_compute_backend_service.default.id
+}
+
+resource "google_compute_target_http_proxy" "default" {
```
replace `default` resource name by something more suitable

loktionovam commented on terraform/lb.tf:
```diff
@@ -0,0 +1,116 @@
+  name    = "http-lb-proxy"
+  url_map = google_compute_url_map.default.id
+}
+resource "google_compute_global_forwarding_rule" "default" {
```
replace `default` resource name by something more suitable

Durf1985 committed:
DO-49 Behavior terraform, add DO-43-49 README

In README file added instructions for all regular tasks and for all tasks
with one "*".

Durf1985 committed:
DO-51 Create several instance via count variables
Created load balancer and instance template
Added in README instruction

loktionovam reviewed (changes requested):
You need to add `terraform/terraform.tfvars.example` because your tests are broken
https://github.com/Durf1985/devops_study_infra/commit/2c26adedd81d4403165630596bf54f05f6b2315d/checks
